simulator/CMakeLists.txt

```diff
@@ -0,0 +1,147 @@
+    infra/string
+    bpu
+    func_sim
+    core
+    mips/mips_register
+    )
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wpedantic -Wold-style-cast -Werror")
+
+if(${DEBUG})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g")
+else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+    set(CMAKE_LD_FLAGS "${CMAKE_CXX_FLAGS} -flto")
+endif()
+
+if(NOT ${RELEASE})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native")
+endif()
+
+if(APPLE) 
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++1z") 
+else() 
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
+endif()
+
+add_executable(${PROJECT_NAME} main.cpp)
+
+foreach(ITER IN LISTS CPPS)
+    get_filename_component(CPP_NAME ${ITER} NAME_WE) 
+    list(APPEND SRCNAMES ${CPP_NAME})
+    get_filename_component(CPP_DIR ${ITER} DIRECTORY)
+    list(APPEND SRCDIRS ${CPP_DIR})
+    add_library(${CPP_NAME} OBJECT ${ITER})
+    target_sources(${PROJECT_NAME} PUBLIC $<TARGET_OBJECTS:${CPP_NAME}>)
+endforeach()
+
+
+#include headers
+include_directories(SRCDIRS ./.)
+
+#include Boost libs
+find_package(Boost COMPONENTS program_options REQUIRED)
+include_directories(${Boost_INCLUDE_DIRS})
+
+#libelf
+find_path(LIBELF_INCLUDE_DIRS
+    NAMES
+        libelf.h
+    PATHS
+        /usr/include
+        /usr/include/libelf
+        /usr/local/include
+        /usr/local/include/libelf
+        /opt/local/include
+        /opt/local/include/libelf
+        /sw/include
+        /sw/include/libelf
+        ENV CPATH)
+
+if(NOT LIBELF_INCLUDE_DIRS)
+    message(SEND_ERROR "Can't find libelf.h")
+endif()
+
+find_library (LIBELF_LIBRARIES
+    NAMES
+        elf
+    PATHS
+        /usr/lib
+        #/usr/local/lib
+        #/opt/local/lib
+        #/sw/lib
+        ENV LIBRARY_PATH
+        ENV LD_LIBRARY_PATH)
+
+if(NOT LIBELF_LIBRARIES)
+    message(SEND_ERROR "Can't find libelf")
+endif()
+
+include_directories(${LIBELF_INCLUDE_DIRS})
+
+target_link_libraries(${PROJECT_NAME} ${Boost_LIBRARIES} ${LIBELF_LIBRARIES})
+
+#clang-tidy
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+###### googletest ######
+add_subdirectory(../googletest/googletest ./build)
+enable_testing()
+add_definitions(-DTEST_PATH=\"${CMAKE_CURRENT_BINARY_DIR}/../../traces\")
+#==================
+foreach(ITER IN LISTS TESTS)
+    #copy all filesfrom */t to build
+    string(CONCAT GLOBBING_EXPR ${ITER} "/t/*.*")
+    file(GLOB INPFILES LIST_DIRECTORIES false ${GLOBBING_EXPR})
+    foreach(ITER2 IN LISTS INPFILES)
+        file(COPY ${ITER2} DESTINATION ./.)
+    endforeach()
+    
+    string(CONCAT SRC_UNIT_TEST ${ITER} "/t/unit_test.cpp")
+    string(REGEX REPLACE "/" "_"  EXEC_NAME_NOTFULL ${ITER})
+    string(CONCAT EXEC_NAME ${EXEC_NAME_NOTFULL} "_test")
+    
+    add_executable(${EXEC_NAME} ${SRC_UNIT_TEST})
+    target_link_libraries(${EXEC_NAME} gtest ${LIBELF_LIBRARIES} ${Boost_LIBRARIES})
+    foreach(ITER2 IN LISTS SRCNAMES)
+        target_sources(${EXEC_NAME} PUBLIC $<TARGET_OBJECTS:${ITER2}>)
```

Comment:
https://cmake.org/pipermail/cmake/2016-June/063749.html

object libraries are not intended for linking. The way it works is you
expand the $<TARGET_OBJECTS> genex into the list of *sources* of the
consuming target. In other words, the string '$<TARGET_OBJECTS:main>' needs
to go into the add_executable(foo) or add_library(foo) command, not into
target_link_libraries(foo).

Reply:
That's strange. Documentation says:
![image](https://user-images.githubusercontent.com/26014259/37049449-f093bfb6-2181-11e8-8e73-3912a39b185c.png)

Reply:
https://cmake.org/cmake/help/v3.1/command/target_sources.html